crates/red_knot_python_semantic/src/types.rs

```diff
@@ -654,14 +654,20 @@ impl<'db> Type<'db> {
                         },
                     )
             }
-            (Type::ClassLiteral(self_class), Type::SubclassOf(target_class)) => {
-                self_class.class.is_subclass_of_base(db, target_class.base)
-            }
+            (
+                Type::ClassLiteral(ClassLiteralType { class: self_class }),
+                Type::SubclassOf(SubclassOfType {
+                    base: ClassBase::Class(target_class),
+                }),
+            ) => self_class.is_subclass_of(db, target_class),
             (
                 Type::Instance(InstanceType { class: self_class }),
-                Type::SubclassOf(target_class),
-            ) if self_class.is_known(db, KnownClass::Type) => {
-                self_class.is_subclass_of_base(db, target_class.base)
+                Type::SubclassOf(SubclassOfType {
+                    base: ClassBase::Class(target_class),
+                }),
+            ) => {
+                self_class.is_known(db, KnownClass::Type)
+                    && target_class.is_known(db, KnownClass::Type)
```

Comment:
Took me a bit to realize that we don't need to handle `SubclassOf(object)` here (like the previous code did), because `Instance(type)` is equivalent to `SubclassOf(object)`, and we check equivalence first in `is_subtype_of`. Might be worth a comment, because (taken on its own) this case now appears incomplete.

TBH I think all these type-algebra methods would be easier to read if every case were commented with a Python-type-syntax example of what the case is checking, but that's out of scope for this PR.

Reply:
Uhhh I actually think this branch is still wrong and no tests fail if I delete it entirely. But I think we do need something here.

`Instance(enum.EnumMeta)` should be a subtype of `type[object]` but not a subtype of `type[type]`. Because instances of `enum.EnumMeta` (example: the class `enum.Enum`) are subclasses of `object` but not subclasses of `type`.

I think I've spent too many hours thinking about `type[]` now because my head's spinning and I can't figure out what logic I need to add to make sure that that works as expected 🙃

Reply:
Yeah, I think this case, as currently written, is wrong: "instances of `type`" is not a subtype of "subclasses of `type`".

> `Instance(enum.EnumMeta)` should be a subtype of `type[object]` but not a subtype of `type[type]`. Because instances of `enum.EnumMeta` (example: the class `enum.Enum`) are subclasses of `object` but not subclasses of `type`.

Yeah, I think the most we can say here is that `Instance(<any subclass of type>)` is a subtype of `SubclassOf(object)`. Which is a generalization from what we have already, which is that `Instance(type)` is equivalent to `SubclassOf(object)`. We probably do need to add a case for this here in `is_subtype_of`.